From 25a28a1f2ec15c174e00c636169b0891dc7f6478 Mon Sep 17 00:00:00 2001
From: VNSL Durga <vnsl.durga.challa@xilinx.com>
Date: Wed, 30 Mar 2016 17:58:48 +0530
Subject: [PATCH 020/827] dma: zynqmp: Added clk unpreparations

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

corrected clk unpreparations based on
error conditions.

Signed-off-by: VNSL Durga <vnsldurg@xilinx.com>
Reviewed-by: Kedareswara rao Appana <appanad@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 712e638e854bcb6bca89a0733e1d33092650b06c)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/dma/xilinx/zynqmp_dma.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/dma/xilinx/zynqmp_dma.c b/drivers/dma/xilinx/zynqmp_dma.c
index 0a2e3ab..09a7d20 100644
--- a/drivers/dma/xilinx/zynqmp_dma.c
+++ b/drivers/dma/xilinx/zynqmp_dma.c
@@ -1045,8 +1045,6 @@ static void zynqmp_dma_chan_remove(struct zynqmp_dma_chan *chan)
 	devm_free_irq(chan->zdev->dev, chan->irq, chan);
 	tasklet_kill(&chan->tasklet);
 	list_del(&chan->common.device_node);
-	clk_disable_unprepare(chan->clk_apb);
-	clk_disable_unprepare(chan->clk_main);
 }
 
 /**
@@ -1149,6 +1147,7 @@ static int zynqmp_dma_chan_probe(struct zynqmp_dma_device *zdev,
 
 	err = clk_prepare_enable(chan->clk_apb);
 	if (err) {
+		clk_disable_unprepare(chan->clk_main);
 		dev_err(&pdev->dev, "Unable to enable apb clock.\n");
 		return err;
 	}
@@ -1224,13 +1223,17 @@ static int zynqmp_dma_probe(struct platform_device *pdev)
 	if (ret) {
 		dev_err(&pdev->dev, "Unable to register DMA to DT\n");
 		dma_async_device_unregister(&zdev->common);
-		goto free_chan_resources;
+		goto free_resources;
 	}
 
 	dev_info(&pdev->dev, "ZynqMP DMA driver Probe success\n");
 
 	return 0;
 
+free_resources:
+	clk_disable_unprepare(zdev->chan->clk_apb);
+	clk_disable_unprepare(zdev->chan->clk_main);
+
 free_chan_resources:
 	zynqmp_dma_chan_remove(zdev->chan);
 	return ret;
@@ -1248,6 +1251,8 @@ static int zynqmp_dma_remove(struct platform_device *pdev)
 
 	of_dma_controller_free(pdev->dev.of_node);
 	dma_async_device_unregister(&zdev->common);
+	clk_disable_unprepare(zdev->chan->clk_apb);
+	clk_disable_unprepare(zdev->chan->clk_main);
 
 	zynqmp_dma_chan_remove(zdev->chan);
 
-- 
2.9.3

